File: firmware/common/src/capability.c
/*
 * Capability report encoding. Roadmap T3.5. Contract in capability.h.
 *
 * Thin by design: the CBOR subset in cbor.c does the encoding, and this file is
 * only the shape of the message and the bounds on what will be accepted back.
 */
#include "lorahome/capability.h"

#include <string.h>

#include "lorahome/cbor.h"
#include "lorahome/protocol.h"

/*
 * The 6-byte layout is a protocol contract, not an implementation detail — the
 * roadmap states it and the host mirrors it. A uint16_t next to four bytes is
 * exactly the arrangement a target's alignment rules can turn into 8, which is
 * why this is asserted on every ABI the compile sweep reaches rather than on
 * the one that happened to run the tests.
 */
_Static_assert(sizeof(lh_capability_t) == 6, "capability drift");
_Static_assert(LH_CAP_MAX == 8, "a report holds one entry per component");

/*
 * The requirement the wire budget is really about: a discovery response must
 * never fragment. Fragmenting it would make discovery depend on reassembly,
 * which depends on an ARQ, and all of that is what you would be debugging when
 * a new node fails to appear.
 *
 * Asserted against the worst case rather than the typical one. A bound that
 * only holds for the reports the encoder happens to produce is not a bound on
 * what a receiver has to survive.
 */
_Static_assert(
    LH_CAP_WIRE_WORST_CASE <= LORAHOME_MAX_PAYLOAD,
    "a capability report must fit one frame without fragmentation");
_Static_assert(LH_CAP_WIRE_BUDGET <= LH_CAP_WIRE_WORST_CASE, "the tighter target is the tighter one");

int lh_cap_encode(const lh_cap_report_t *report, uint8_t *out, uint16_t cap) {
  if (report == NULL || report->count > LH_CAP_MAX) return -1;

  lh_cbor_writer_t writer;
  lh_cbor_writer_init(&writer, out, cap);

  lh_cbor_write_map(&writer, 3);

  lh_cbor_write_uint(&writer, LH_CAP_KEY_FW_VERSION);
  lh_cbor_write_uint(&writer, report->fw_version);

  lh_cbor_write_uint(&writer, LH_CAP_KEY_FREE_HEAP_KB);
  lh_cbor_write_uint(&writer, report->free_heap_kb);

  lh_cbor_write_uint(&writer, LH_CAP_KEY_COMPONENTS);
  lh_cbor_write_array(&writer, report->count);

  for (uint8_t i = 0; i < report->count; i++) {
    const lh_capability_t *entry = &report->caps[i];
    lh_cbor_write_array(&writer, LH_CAP_FIELDS);
    lh_cbor_write_uint(&writer, entry->driver_type_id);
    lh_cbor_write_uint(&writer, entry->bus_addr);
    lh_cbor_write_uint(&writer, entry->bus_type);
    lh_cbor_write_uint(&writer, entry->channel_count);
    lh_cbor_write_uint(&writer, entry->flags);
  }

  /* One check at the end. The writer stops emitting once it has overflowed, so
   * a partial buffer can never be mistaken for a short message. */
  return lh_cbor_writer_finish(&writer);
}

/** Reads one component array, tolerating extra trailing fields. */
static bool decode_entry(lh_cbor_reader_t *reader, lh_capability_t *out) {
  uint32_t fields = 0;
  if (!lh_cbor_read_array(reader, &fields)) return false;
  if (fields < LH_CAP_FIELDS) return false;

  uint64_t values[LH_CAP_FIELDS];
  for (uint8_t i = 0; i < LH_CAP_FIELDS; i++) {
    if (!lh_cbor_read_uint(reader, &values[i])) return false;
  }
  /* A newer node may describe a component with more fields than this build
   * knows about. Skipping them is what lets an old host still discover it,
   * rather than reporting a device it can see as unreadable. */
  for (uint32_t i = LH_CAP_FIELDS; i < fields; i++) {
    if (!lh_cbor_skip(reader)) return false;
  }

  if (values[LH_CAP_FIELD_TYPE_ID] > UINT16_MAX) return false;
  if (values[LH_CAP_FIELD_BUS_ADDR] > UINT8_MAX) return false;
  if (values[LH_CAP_FIELD_BUS_TYPE] > UINT8_MAX) return false;
  if (values[LH_CAP_FIELD_CHANNELS] > UINT8_MAX) return false;
  if (values[LH_CAP_FIELD_FLAGS] > UINT8_MAX) return false;

  out->driver_type_id = (uint16_t)values[LH_CAP_FIELD_TYPE_ID];
  out->bus_addr = (uint8_t)values[LH_CAP_FIELD_BUS_ADDR];
  out->bus_type = (uint8_t)values[LH_CAP_FIELD_BUS_TYPE];
  out->channel_count = (uint8_t)values[LH_CAP_FIELD_CHANNELS];
  out->flags = (uint8_t)values[LH_CAP_FIELD_FLAGS];
  return true;
}

bool lh_cap_decode(const uint8_t *buf, uint16_t len, lh_cap_report_t *out) {
  if (out == NULL) return false;
  memset(out, 0, sizeof *out);

  lh_cbor_reader_t reader;
  lh_cbor_reader_init(&reader, buf, len);

  uint32_t pairs = 0;
  if (!lh_cbor_read_map(&reader, &pairs)) return false;

  for (uint32_t pair = 0; pair < pairs; pair++) {
    uint64_t key = 0;
    if (!lh_cbor_read_uint(&reader, &key)) return false;

    switch (key) {
      case LH_CAP_KEY_FW_VERSION: {
        uint64_t value = 0;
        if (!lh_cbor_read_uint(&reader, &value)) return false;
        if (value > UINT32_MAX) return false;
        out->fw_version = (uint32_t)value;
        break;
      }
      case LH_CAP_KEY_FREE_HEAP_KB: {
        uint64_t value = 0;
        if (!lh_cbor_read_uint(&reader, &value)) return false;
        if (value > UINT16_MAX) return false;
        out->free_heap_kb = (uint16_t)value;
        break;
      }
      case LH_CAP_KEY_COMPONENTS: {
        uint32_t count = 0;
        if (!lh_cbor_read_array(&reader, &count)) return false;
        /* A node that reports more components than this build can hold is a
         * protocol mismatch, not something to silently truncate: a host acting
         * on the first eight of twelve would show a device with parts missing
         * and no indication that anything was dropped. */
        if (count > LH_CAP_MAX) return false;
        for (uint32_t i = 0; i < count; i++) {
          if (!decode_entry(&reader, &out->caps[i])) return false;
        }
        out->count = (uint8_t)count;
        break;
      }
      default:
        /* Forward compatibility: a key from a newer protocol is stepped over,
         * so adding one does not require flashing every node in a building on
         * the same day. */
        if (!lh_cbor_skip(&reader)) return false;
        break;
    }
  }

  return !reader.error;
}

bool lh_cap_equal(const lh_cap_report_t *a, const lh_cap_report_t *b) {
  if (a == NULL || b == NULL) return false;
  if (a->count != b->count) return false;
  if (a->fw_version != b->fw_version) return false;
  if (a->free_heap_kb != b->free_heap_kb) return false;
  /* Field by field rather than memcmp over the struct: padding bytes are
   * whatever the last write left there, and comparing them would make the
   * answer depend on how the report was built. */
  for (uint8_t i = 0; i < a->count; i++) {
    if (a->caps[i].driver_type_id != b->caps[i].driver_type_id) return false;
    if (a->caps[i].bus_addr != b->caps[i].bus_addr) return false;
    if (a->caps[i].bus_type != b->caps[i].bus_type) return false;
    if (a->caps[i].channel_count != b->caps[i].channel_count) return false;
    if (a->caps[i].flags != b->caps[i].flags) return false;
  }
  return true;
}

File: packages/protocol/src/capability.ts
/**
 * Capability discovery. Twin of firmware/common/src/capability.c (T3.5).
 *
 * The Host asks with a 0x40 frame; the node answers with 0x41 carrying this
 * report. It inverts the usual flow — rather than declaring a sensor in a
 * configuration file and hoping it is wired correctly, you plug the hardware in
 * and the system tells you what it found (ARCHITECTURE.md §8).
 *
 * The report must fit one frame. Fragmenting a discovery response would make
 * discovery depend on reassembly, which depends on an ARQ — a lot of machinery
 * to answer "what are you?", and all of it is what you would be debugging when
 * a new node fails to appear.
 */

import { CborReader, CborWriter } from './cbor.js';

/** Components in a report. The node's own limit, so a report is always complete. */
export const CAP_MAX = 8;

/**
 * Wire budget for a realistic full report. Checked in the tests, not assumed.
 *
 * Applies to eight components with the field values the protocol actually
 * produces: bus types from a five-value enum, channel counts in single digits,
 * one flag bit defined. It is what constrains adding drivers.
 */
export const CAP_WIRE_BUDGET = 100;

/**
 * The largest report this codec can produce.
 *
 * 12 B per component (1 array header, 3 for a 16-bit type id, 2 each for four
 * bytes that spill past the inline form) plus 13 B of envelope. Still fits one
 * frame with more than half the payload spare, which is the requirement the
 * budget above exists to express.
 */
export const CAP_WIRE_WORST_CASE = 109;

export enum BusType {
  I2C = 0,
  GPIO = 1,
  SPI = 2,
  OneWire = 3,
  ADC = 4,
}

/** bit 0: this component already has a configuration on the node. */
export const CAP_FLAG_CONFIGURED = 0x01;

/** Map keys. Integers, never strings — CONTRIBUTING.md §3. */
export const CapKey = {
  FwVersion: 1,
  FreeHeapKb: 2,
  Components: 3,
} as const;

/** Positions inside one component's array. */
export const CAP_FIELDS = 5;

export interface Capability {
  driverTypeId: number;
  busAddr: number;
  busType: BusType;
  channelCount: number;
  flags: number;
}

export interface CapabilityReport {
  fwVersion: number;
  freeHeapKb: number;
  components: Capability[];
}

/**
 * Encodes a report.
 *
 * Each component is a positional array rather than a map. A map per component
 * would cost five key bytes each — forty bytes over eight components, against a
 * hundred-byte budget — to describe a structure the protocol fixes and which
 * could not be reordered anyway. The outer report is a map because it *will*
 * gain keys; the components array is exactly where it must not.
 */
export function encodeCapabilityReport(report: CapabilityReport): Uint8Array {
  if (report.components.length > CAP_MAX) {
    throw new RangeError(`a report holds at most ${CAP_MAX} components`);
  }

  const writer = new CborWriter();
  writer.map(3);

  writer.uint(CapKey.FwVersion).uint(report.fwVersion);
  writer.uint(CapKey.FreeHeapKb).uint(report.freeHeapKb);

  writer.uint(CapKey.Components).array(report.components.length);
  for (const component of report.components) {
    writer.array(CAP_FIELDS);
    writer.uint(component.driverTypeId);
    writer.uint(component.busAddr);
    writer.uint(component.busType);
    writer.uint(component.channelCount);
    writer.uint(component.flags);
  }

  return writer.finish();
}

/**
 * Decodes a report.
 *
 * Unknown map keys are skipped rather than refused, so a host can still
 * discover a node running firmware it does not recognise. A component array
 * with more fields than this build knows is truncated the same way, for the
 * same reason. Throws only for a message that is malformed or exceeds the
 * receiver's limits.
 */
export function decodeCapabilityReport(buf: Uint8Array): CapabilityReport {
  const reader = new CborReader(buf);
  const report: CapabilityReport = { fwVersion: 0, freeHeapKb: 0, components: [] };

  const pairs = reader.map();
  for (let pair = 0; pair < pairs; pair++) {
    const key = reader.uint();

    switch (key) {
      case CapKey.FwVersion:
        report.fwVersion = reader.uint();
        break;
      case CapKey.FreeHeapKb:
        report.freeHeapKb = reader.uint();
        break;
      case CapKey.Components: {
        const count = reader.array();
        // More components than this build can hold is a protocol mismatch, not
        // something to truncate: a host acting on the first eight of twelve
        // would show a device with parts missing and no sign anything was lost.
        if (count > CAP_MAX) throw new RangeError(`report claims ${count} components`);
        for (let i = 0; i < count; i++) report.components.push(decodeComponent(reader));
        break;
      }
      default:
        // Forward compatibility: a key from a newer protocol is stepped over,
        // so adding one does not require flashing every node on the same day.
        reader.skip();
        break;
    }
  }

  return report;
}

function decodeComponent(reader: CborReader): Capability {
  const fields = reader.array();
  if (fields < CAP_FIELDS) throw new RangeError(`component has only ${fields} fields`);

  const component: Capability = {
    driverTypeId: reader.uint(),
    busAddr: reader.uint(),
    busType: reader.uint() as BusType,
    channelCount: reader.uint(),
    flags: reader.uint(),
  };

  // A newer node may describe a component with more fields than this build
  // knows. Skipping them lets an old host still discover it, rather than
  // reporting a device it can see as unreadable.
  for (let i = CAP_FIELDS; i < fields; i++) reader.skip();

  return component;
}

/** True if two reports carry identical contents. For tests and diagnostics. */
export function capabilityReportsEqual(a: CapabilityReport, b: CapabilityReport): boolean {
  if (a.fwVersion !== b.fwVersion) return false;
  if (a.freeHeapKb !== b.freeHeapKb) return false;
  if (a.components.length !== b.components.length) return false;

  return a.components.every((component, index) => {
    const other = b.components[index]!;
    return (
      component.driverTypeId === other.driverTypeId &&
      component.busAddr === other.busAddr &&
      component.busType === other.busType &&
      component.channelCount === other.channelCount &&
      component.flags === other.flags
    );
  });
}
